src/oracle/monitor.js

```diff
@@ -0,0 +1,429 @@
+    labelNames: ['feed']
+})
+
+// Register the gaugex
+register.registerMetric(oracleSubmission)
+register.registerMetric(oracleObservation)
+register.registerMetric(oracleLastRound)
+register.registerMetric(oracleBalance)
+register.registerMetric(oracleDeviation)
+register.registerMetric(actualPriceGauge)
+
+const { agoricNames, fromBoard, vstorage } = await makeRpcUtils({ fetch });
+
+//this holds the offer ids
+var feeds = []
+//this holds the amounts in
+var amountsIn = {}
+
+/**
+ * Function to read oracles
+ * @returns oracles, their names and their addresses
+ */
+const readOracleAddresses = () => {
+    var oracles = readJSONFile(ORACLE_FILE)
+    return oracles
+}
+
+/**
+ * Function to get oracles feed invitations
+ */
+export const getOraclesInvitations = async () => {
+    //get the feeds
+    feeds = agoricNames.reverse
+
+    //for each oracle
+    for (let oracle in oracles) {
+
+        const current = await getCurrent(oracle, fromBoard, { vstorage });
+        const invitations = current.offerToUsedInvitation
+
+        //for each invitation
+        for (let inv in invitations) {
+            let boardId = invitations[inv].value[0].instance.boardId
+            let feed = feeds[boardId].split(" price feed")[0]
+
+            if (!("feeds" in oracles[oracle])) {
+                oracles[oracle]["feeds"] = {}
+            }
+            //add feed
+            oracles[oracle]["feeds"][String(inv)] = feed
+        }
+    }
+}
+
+//var oracleLabels = readOracles();
+var oracles = readOracleAddresses();
+await getOraclesInvitations();
+
+/**
+ * Function to update metrics
+ * @param {*} oracleName oracle name
+ * @param {*} oracle oracle address
+ * @param {*} feed feed name
+ * @param {*} value new feed value
+ * @param {*} id submission id which is a timestamp
+ * @param {*} actualPrice feed actual aggregated price
+ * @param {*} lastRound latest round id for which there was a submission
+ */
+const updateMetrics = (oracleName, oracle, feed, value, id, actualPrice, lastRound) => {
+    //calculate price deviation from actual value
+    let priceDeviation = Math.abs((value - actualPrice) / actualPrice) * 100
+
+    oracleSubmission.labels(oracleName, oracle, feed).set(value)
+    oracleObservation.labels(oracleName, oracle, feed).set(id)
+    oracleLastRound.labels(oracleName, oracle, feed).set(lastRound)
+    oracleDeviation.labels(oracleName, oracle, feed).set(priceDeviation)
+    actualPriceGauge.labels(feed).set(actualPrice)
+}
+
+/**
+ * Function to update balance metrics
+ * @param {*} oracleName oracle name
+ * @param {*} oracle oracle address
+ * @param {*} brand brand
+ * @param {*} value balance value to set
+ */
+const updateBalanceMetrics = (oracleName, oracle, brand, value) => {
+    oracleBalance.labels(oracleName, oracle, brand).set(value)
+}
+
+/**
+ * Function to query price for feed
+ * @param {*} feed feed like 'BRAND_IN-BRAND_OUT'
+ * @returns the price of the feed
+ */
+const queryPrice = async (feed) => {
```

Comment:
Code duplication, 99% of this function is already defined in `middleware.js`. I strongly suggest putting one general implementation of these functions in a utils script so you can use just one version throughout. Then you can add specifics in the specific scripts.

Reply:
Note, this suggestion applies to all other functions in this script which are similar.

Reply:
Fixed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/bafa5748c9af7238d24b465797fdb075b783a8b0